src/GNDStk/Tree/src/read.hpp

```diff
@@ -0,0 +1,195 @@
+
+   // fallback: try automagick
+   return read(file,format::null);
+}
+
+
+
+// ------------------------
+// read(istream,format)
+// ------------------------
+
+std::istream &read(
+   std::istream &is,
+   format form = format::null
+) {
+   // clear
+   // Comment as in read(string). Note that we need the clear() here, too,
+   // because this function might be called directly, not via read(string).
+   clear();
+
+   // ------------------------
+   // format::tree
+   // Not allowed in read
+   // ------------------------
+
+   if (form == format::tree) {
+      error(detail::format_tree_read);
+      return is;
+   }
+
+   // ------------------------
+   // format::xml,json,hdf5
+   // Check: consistent
+   // initial content?
+   // ------------------------
+
+   if (is.peek() == '<') {
+      // looks like xml
+      if (form == format::null)
+         form = format::xml;
+      else if (form != format::xml)
+         detail::warning_tree_io_data<dummy>(
+            form, "'<'", "XML");
+   } else if (is.peek() == 137) {
+      // looks like hdf5
+      if (form == format::null)
+         form = format::hdf5;
+      else if (form != format::hdf5)
+         detail::warning_tree_io_data<dummy>(
+            form, "char 137", "HDF5");
+   } else {
+      // looks like json (via process of elimination)
+      if (form == format::null)
+         form = format::json;
+      else if (form != format::json)
+         detail::warning_tree_io_data<dummy>(
+            form, "neither '<' nor char 137", "Json");
+   }
+
+   // The above logic is such that form cannot henceforth be format::tree,
+   // which would have triggered a return, or format::null, which would have
+   // become one of {xml,json,hdf5} somewhere in the above conditional.
+
+   // ------------------------
+   // Read and convert
+   // ------------------------
+
+   // Obey form, independent of whatever might or might not have been warned
+   // about above. Note that if the original form parameter was format::null,
+   // then form would have been modified, above, to the likely correct format,
+   // based on the peek() calls.
+   if (form == format::xml) {
+      // assume .xml
+      // go through a temporary xml object to create the tree...
+      const xml tmp(is);
+      if (not is.fail())
+         convert(tmp, *this);
+   } else if (form == format::json) {
+      // assume .json
+      // go through a temporary json object to create the tree...
+      const json tmp(is);
+
+      // It would seem that the nlohmann::json stream input operation,
+      // which is used by the constructor we just called, sets failbit
+      // in instances in which it should just set eofbit. So, for now,
+      // we'll comment-out the "not is.fail()" test... :-/
+      // if (not is.fail())
+      convert(tmp, *this);
+   } else if (form == format::hdf5) {
+      error("HDF5 read() is not implemented yet");
+   } else {
+      // we may or may not want slippery-slope stuff like this
```

Comment:
We definitely want to fall off like this. If we can't recognize the form, we can't really do anything. Elsewhere we have thrown exceptions and have expected the caller to catch it if it needs.

I know throwing exceptions is very costly, but if you get into a situation where you are throwing/catching exceptions, your performance doesn't matter anymore.

Reply:
In this particular case, it turns out that the structure of an earlier conditional was such that it's impossible for the "else" in question to happen, ever. Unless, I suppose, a random cosmic ray hits the computer and messes with the electronics. This was why I wrote the "slippery-slope" remark: because I'm really checking something that would be a 100% internal error in the code. Something that, realistically (so, chuck the cosmic ray business), would be triggered only if someone went into the earlier code, sometime in the future, and somehow changed it in such a way that they broke it. Filling code with "double checking" like this would be a slippery slope, and could make virtually any development on any software effectively grind to a halt.